assets/src/edit-story/app/story/actions/useSetPropertiesById.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * WordPress dependencies
+ */
+import { useCallback } from '@wordpress/element';
+
+/**
+ * Sets element properties by ID.
+ *
+ * @param {number}   currentPageIndex Index of the page currently in focus.
+ * @param {Array}    pages Array of all pages.
+ * @param {Function} setPages Setting the state of pages.
+ * @return {Function} Sets properties by element Id.
+ */
+function useSetPropertiesById( {
+	currentPageIndex,
+	pages,
+	setPages,
+} ) {
+	const setPropertiesById = useCallback( ( elementId, properties ) => {
+		const currentPage = pages[ currentPageIndex ];
+		const newPages = [
+			...pages.slice( 0, currentPageIndex ),
+			{
+				...currentPage,
+				elements: [
+					...currentPage.elements.map( ( element ) => {
+						const { id } = element;
+						// If ids don't match, let's return the element as it was.
+						if ( elementId !== id ) {
+							return element;
+						}
+						return {
+							...element,
+							...properties,
+						};
+					} ),
+				],
+			},
+			...pages.slice( currentPageIndex + 1 ),
+		];
+		setPages( newPages );
```

Comment:
This probably needs to use the functional setter to allow multiple elements being updated in the same render loop:

```js
setPages( ( pages ) => {
	// create newPages
	return newPages;
} );
```

That'll also completely remove the need for passing `pages` in at all and just use it inline in the setter.

Actually, all the methods ought to do this now that I think about it.

Reply:
Should we perhaps generally have smaller exposed actions that are using non-exposed internal logic within the `useStory`?

E.g. we could have `setPropertiesById` and `setPropertiesByIds` both exposed via `useStory`, however, internally this would always use `setPropertiesByIds` anyway. Perhaps even `setPropertiesBy` could be a general internal function which would accept `by` (e.g. selection or ids)?

Basically just to have the base logic of setting the properties via slicing the `pages` in one place. Thoughts?

Reply:
> Basically just to have the base logic of setting the properties via slicing the `pages` in one place. Thoughts?

We could easily do this where relevant. Perhaps it's even better to do it with a reducer and simple actions (but expose it as functions) to move the logic to a more centralized place, as suggested by @dvoytenko elsewhere.

We could probably keep it to the following actions in the reducer and re-use those for the different exposed functions:

```
PAGE_DELETE
PAGE_ADD
PAGE_UPDATE
ELEMENT_DELETE
ELEMENT_ADD
ELEMENT_UPDATE
```

All of the exposed functions would then be a series of reductions of the same state. But can we actually easily do a series of reductions?

Or we would create reducers for each atomic action - we'd need 4 extra actions for duplicating both pages and elements, updating multiple elements and deleting multiple elements. Seems doable too.

And we could get all of the state variables such as currentPage, currentPageId, currentPageNumber, etc directly from the state.

It's probably the right way to do it.

I'll create a (high-priority) ticket about it.

Reply:
I added this ticket a bit ago: https://github.com/ampproject/amp-wp/issues/3894

Reply:
@barklund On updating multiple elements in a loop: decided to go with [updateElementsByIds action](https://github.com/ampproject/amp-wp/pull/3909/files#diff-368f816669cc6567ef5ec19001aa7210) instead which would work for both single and multiple element updates. It's currently in the multi-selection PR only, however, will replace the one within this PR with that one. Let me know if you have any concerns.

Reply:
All this will be irrelevant when #3910 is merged in.